### PR TITLE
Account trimAboveSeqNo in committed translog generation

### PR DESCRIPTION
**Pull request by dnhatn**
While working on the fix for #49970, I found that we do not consider trimAboveSeqNo when calculating the translog generation of an index commit. If there is no new indexing after the primary promotion, then we won't be able to clean up the translog.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/translog/Translog.java [hunk 1]
```diff
@@ -697,10 +697,7 @@ private Snapshot newMultiSnapshot(TranslogSnapshot[] snapshots) throws IOExcepti
             "callers of readersAboveMinSeqNo must hold a lock: readLock ["
                 + readLock.isHeldByCurrentThread() + "], writeLock [" + readLock.isHeldByCurrentThread() + "]";
         return Stream.concat(readers.stream(), Stream.of(current))
-            .filter(reader -> {
-                final long maxSeqNo = reader.getCheckpoint().maxSeqNo;
-                return maxSeqNo == SequenceNumbers.UNASSIGNED_SEQ_NO || maxSeqNo >= minSeqNo;
-            });
+            .filter(reader -> minSeqNo <= SequenceNumbers.min(reader.getCheckpoint().trimmedAboveSeqNo, reader.getCheckpoint().maxSeqNo));
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/index/translog/Translog.java]: I think we need to account for the case where trimmedAboveSeqNo is equal to `SequenceNumbers.UNASSIGNED_SEQ_NO`. Perhaps we can add a method to checkpoint called `getMaxEffectiveSeqNo` which is:
```
if (trimmedAboveSeqNo == SequenceNumbers.UNASSIGNED_SEQ_NO)
  return maxSeqNo;
else
  return Math.min(maxSeqNo, trimmedAboveSeqNo);
```

**Review by ywelsch — CHANGES_REQUESTED** [review]

File: server/src/main/java/org/elasticsearch/index/translog/Translog.java (hunk 1 above)
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/index/translog/Translog.java]: `SequenceNumbers.min` should handle this case:
https://github.com/elastic/elasticsearch/blob/044139a8432f9d14190de365402f0a0da1212b18/server/src/main/java/org/elasticsearch/index/seqno/SequenceNumbers.java#L84-L85

Having an explicit method is better. I pushed https://github.com/elastic/elasticsearch/pull/50205/commits/563545c5ee3629f5e2816463020815716aa65e17.

Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/index/translog/Translog.java]: Ah lol, I did not see that a special `min` method was used (I read it as Math.min).

**Review by ywelsch — APPROVED** [review]
LGTM

**Comment by dnhatn** [comment]
Thanks Yannick.

**Pull request merged**